File: infra/communication/python/pacific_rim_communication_infra/core/routing.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Mapping

from pacific_rim_communication_infra.contracts import Endpoint, MiddlewareConfig, PubSubRoute, RpcRoute, TransportKind

from .middleware import Channel, MessageBus, channel_from_endpoint, create_message_bus_from_config
from .security import SecureMessageBus, SecurityRuntime


@dataclass(frozen=True)
class BoundEndpoint:
  bus_name: str
  bus: MessageBus
  channel: Channel


class CommunicationFabric:
  def __init__(
    self,
    buses: Mapping[str, MessageBus],
    bus_configs: Mapping[str, MiddlewareConfig],
    pubsub_routes: Iterable[PubSubRoute] = (),
    rpc_routes: Iterable[RpcRoute] = (),
    security: SecurityRuntime | None = None,
  ):
    self._buses = dict(buses)
    self._bus_configs = dict(bus_configs)
    self._pubsub_routes = {route.name: route for route in pubsub_routes if route.enabled}
    self._rpc_routes = {route.name: route for route in rpc_routes if route.enabled}
    self._security = security

  @classmethod
  def from_configs(
    cls,
    middleware: Mapping[str, MiddlewareConfig],
    pubsub_routes: Iterable[PubSubRoute] = (),
    rpc_routes: Iterable[RpcRoute] = (),
    security: SecurityRuntime | None = None,
  ) -> "CommunicationFabric":
    buses = {name: create_message_bus_from_config(config) for name, config in middleware.items()}
    return cls(buses, middleware, pubsub_routes, rpc_routes, security=security)

  async def connect_all(self) -> None:
    for bus in self._buses.values():
      await bus.connect()

  async def close_all(self) -> None:
    for bus in self._buses.values():
      await bus.close()

  def bus(self, name: str) -> MessageBus:
    return self._buses[name]

  def publisher(self, route_name: str) -> BoundEndpoint:
    route = self._pubsub_routes[route_name]
    return self._bind_endpoint(route.publisher)

  def subscriber(self, route_name: str) -> BoundEndpoint:
    route = self._pubsub_routes[route_name]
    return self._bind_endpoint(route.subscriber)

  def rpc_client(self, route_name: str) -> BoundEndpoint:
    route = self._rpc_routes[route_name]
    bus_name, bus = self._bus_for_endpoint(route.client)
    bus = self._secure_bus_for_endpoint(bus_name, bus, route.server)
    channel = channel_from_endpoint(
      route.server.address,
      message_type=route.server.message_type,
      queue_group=str(route.server.metadata.get("queue_group", "")),
      metadata=route.server.metadata,
    )
    return BoundEndpoint(bus_name=bus_name, bus=bus, channel=channel)

  def rpc_server(self, route_name: str) -> BoundEndpoint:
    route = self._rpc_routes[route_name]
    return self._bind_endpoint(route.server)

  def _bind_endpoint(self, endpoint: Endpoint) -> BoundEndpoint:
    bus_name, bus = self._bus_for_endpoint(endpoint)
    bus = self._secure_bus_for_endpoint(bus_name, bus, endpoint)
    channel = channel_from_endpoint(
      endpoint.address,
      message_type=endpoint.message_type,
      queue_group=str(endpoint.metadata.get("queue_group", "")),
      metadata=endpoint.metadata,
    )
    return BoundEndpoint(bus_name=bus_name, bus=bus, channel=channel)

  def _bus_for_endpoint(self, endpoint: Endpoint) -> tuple[str, MessageBus]:
    explicit_name = endpoint.metadata.get("middleware.runtime") or endpoint.metadata.get("middleware_name") or endpoint.metadata.get("middleware")
    if explicit_name:
      return str(explicit_name), self._buses[str(explicit_name)]

    for name, config in self._bus_configs.items():
      if config.transport == endpoint.transport:
        return name, self._buses[name]

    raise KeyError(f"no middleware configured for transport {endpoint.transport}")

  def _secure_bus_for_endpoint(self, bus_name: str, bus: MessageBus, endpoint: Endpoint) -> MessageBus:
    if self._security is None:
      return bus
    binding = self._security.resolve_binding(bus_name, self._bus_configs[bus_name], endpoint)
    if binding is None:
      return bus
    return SecureMessageBus(bus, binding)


def default_middleware_name(transport: TransportKind) -> str:
  return transport.value

File: infra/communication/go/core/routing.go
package core

import (
	"context"
	"fmt"

	communication "github.com/pacific-rim/pacific-rim/infra/communication/go/contracts"
)

type BoundEndpoint struct {
	BusName string
	Bus     MessageBus
	Channel Channel
}

type Fabric struct {
	buses        map[string]MessageBus
	busConfigs   map[string]BusConfig
	pubsubRoutes map[string]communication.PubSubRoute
	rpcRoutes    map[string]communication.RPCRoute
	security     *SecurityRuntime
}

func NewFabric(
	busConfigs map[string]BusConfig,
	pubsubRoutes []communication.PubSubRoute,
	rpcRoutes []communication.RPCRoute,
) (*Fabric, error) {
	return NewFabricWithSecurity(busConfigs, pubsubRoutes, rpcRoutes, nil)
}

func NewFabricWithSecurity(
	busConfigs map[string]BusConfig,
	pubsubRoutes []communication.PubSubRoute,
	rpcRoutes []communication.RPCRoute,
	security *SecurityRuntime,
) (*Fabric, error) {
	buses := make(map[string]MessageBus, len(busConfigs))
	for name, config := range busConfigs {
		bus, err := NewBus(config)
		if err != nil {
			return nil, err
		}
		buses[name] = bus
	}

	fabric := &Fabric{
		buses:        buses,
		busConfigs:   busConfigs,
		pubsubRoutes: map[string]communication.PubSubRoute{},
		rpcRoutes:    map[string]communication.RPCRoute{},
		security:     security,
	}
	for _, route := range pubsubRoutes {
		if route.Enabled {
			fabric.pubsubRoutes[route.Name] = route
		}
	}
	for _, route := range rpcRoutes {
		if route.Enabled {
			fabric.rpcRoutes[route.Name] = route
		}
	}
	return fabric, nil
}

func (f *Fabric) ConnectAll(ctx context.Context) error {
	for _, bus := range f.buses {
		if err := bus.Connect(ctx); err != nil {
			return err
		}
	}
	return nil
}

func (f *Fabric) CloseAll(ctx context.Context) error {
	for _, bus := range f.buses {
		if err := bus.Close(ctx); err != nil {
			return err
		}
	}
	return nil
}

func (f *Fabric) Bus(name string) (MessageBus, bool) {
	if f == nil {
		return nil, false
	}
	bus, ok := f.buses[name]
	return bus, ok
}

func (f *Fabric) FirstBusForTransport(kind communication.TransportKind) (string, MessageBus, bool) {
	if f == nil {
		return "", nil, false
	}
	for name, config := range f.busConfigs {
		if config.Transport == kind {
			bus, ok := f.buses[name]
			if ok {
				return name, bus, true
			}
		}
	}
	return "", nil, false
}

func (f *Fabric) Publisher(routeName string) (BoundEndpoint, error) {
	route, ok := f.pubsubRoutes[routeName]
	if !ok {
		return BoundEndpoint{}, fmt.Errorf("pubsub route %q is not configured", routeName)
	}
	return f.bindEndpoint(route.Publisher)
}

func (f *Fabric) Subscriber(routeName string) (BoundEndpoint, error) {
	route, ok := f.pubsubRoutes[routeName]
	if !ok {
		return BoundEndpoint{}, fmt.Errorf("pubsub route %q is not configured", routeName)
	}
	return f.bindEndpoint(route.Subscriber)
}

func (f *Fabric) RPCClient(routeName string) (BoundEndpoint, error) {
	route, ok := f.rpcRoutes[routeName]
	if !ok {
		return BoundEndpoint{}, fmt.Errorf("rpc route %q is not configured", routeName)
	}
	busName, bus, err := f.busForEndpoint(route.Client)
	if err != nil {
		return BoundEndpoint{}, err
	}
	bus, err = f.secureBusForEndpoint(busName, bus, route.Server)
	if err != nil {
		return BoundEndpoint{}, err
	}
	return BoundEndpoint{
		BusName: busName,
		Bus:     bus,
		Channel: ChannelFromEndpoint(route.Server),
	}, nil
}

func (f *Fabric) RPCServer(routeName string) (BoundEndpoint, error) {
	route, ok := f.rpcRoutes[routeName]
	if !ok {
		return BoundEndpoint{}, fmt.Errorf("rpc route %q is not configured", routeName)
	}
	return f.bindEndpoint(route.Server)
}

func (f *Fabric) bindEndpoint(endpoint communication.Endpoint) (BoundEndpoint, error) {
	busName, bus, err := f.busForEndpoint(endpoint)
	if err != nil {
		return BoundEndpoint{}, err
	}
	bus, err = f.secureBusForEndpoint(busName, bus, endpoint)
	if err != nil {
		return BoundEndpoint{}, err
	}
	return BoundEndpoint{
		BusName: busName,
		Bus:     bus,
		Channel: ChannelFromEndpoint(endpoint),
	}, nil
}

func (f *Fabric) secureBusForEndpoint(busName string, bus MessageBus, endpoint communication.Endpoint) (MessageBus, error) {
	if f == nil || f.security == nil {
		return bus, nil
	}
	config, ok := f.busConfigs[busName]
	if !ok {
		return bus, nil
	}
	binding, err := f.security.ResolveBinding(busName, config, endpoint)
	if err != nil {
		return nil, err
	}
	if binding == nil {
		return bus, nil
	}
	return NewSecureMessageBus(bus, *binding)
}

func (f *Fabric) busForEndpoint(endpoint communication.Endpoint) (string, MessageBus, error) {
	if endpoint.Metadata != nil {
		if name := firstEndpointMetadata(endpoint.Metadata, "middleware.runtime", "middleware_name", "middleware"); name != "" {
			bus, ok := f.buses[name]
			if !ok {
				return "", nil, fmt.Errorf("middleware %q is not configured", name)
			}
			return name, bus, nil
		}
	}

	for name, config := range f.busConfigs {
		if config.Transport == endpoint.Transport {
			return name, f.buses[name], nil
		}
	}
	return "", nil, fmt.Errorf("no middleware configured for transport %q", endpoint.Transport)
}

func firstEndpointMetadata(metadata map[string]string, keys ...string) string {
	for _, key := range keys {
		if value := metadata[key]; value != "" {
			return value
		}
	}
	return ""
}

func CanonicalRouteName(value string) string {
	return routeName(value)
}
